<a href="https://colab.research.google.com/github/vaguiarfinance/AprendizadoDNC/blob/main/Variaveis_Dummy.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [5]:
import pandas as pd
import numpy as np
import seaborn as sns

In [2]:
limite = pd.read_excel("Limite_Credito.xlsx")
limite.head()

Idade  RendimentoTotal  Salario  LimitedeCreditoImediato  \
0     70            10736     5214                      400   
1     68             5000     5000                      380   
2     44             5800     5800                      500   
3     72             4300     4300                     2000   
4     75             4400     4400                     3000   

   LimitedoChequeEspecial       Escolaridade  
0                     500  Fundamental_Medio  
1                     600  Fundamental_Medio  
2                     800  Fundamental_Medio  
3                    1000  Fundamental_Medio  
4                    1000  Fundamental_Medio

In [4]:
# Já que o modelo não realiza a leitura variáveis categoricas, por isso vamos transformar para Dummys

limite = pd.get_dummies(limite, prefix=["Escolaridade"])
limite.head()

Idade  RendimentoTotal  Salario  LimitedeCreditoImediato  \
0     70            10736     5214                      400   
1     68             5000     5000                      380   
2     44             5800     5800                      500   
3     72             4300     4300                     2000   
4     75             4400     4400                     3000   

   LimitedoChequeEspecial  Escolaridade_Fundamental_Medio  \
0                     500                               1   
1                     600                               1   
2                     800                               1   
3                    1000                               1   
4                    1000                               1   

   Escolaridade_Superior_Pos  
0                          0  
1                          0  
2                          0  
3                          0  
4                          0

In [5]:
# Vamos tirar a correlação entre as variáveis

limite.corr()

Idade  RendimentoTotal   Salario  \
Idade                           1.000000         0.066396  0.063227   
RendimentoTotal                 0.066396         1.000000  0.960551   
Salario                         0.063227         0.960551  1.000000   
LimitedeCreditoImediato         0.064149         0.031041  0.081792   
LimitedoChequeEspecial          0.075883         0.858120  0.888839   
Escolaridade_Fundamental_Medio -0.110892        -0.396191 -0.475479   
Escolaridade_Superior_Pos       0.110892         0.396191  0.475479   

                                LimitedeCreditoImediato  \
Idade                                          0.064149   
RendimentoTotal                                0.031041   
Salario                                        0.081792   
LimitedeCreditoImediato                        1.000000   
LimitedoChequeEspecial                         0.226990   
Escolaridade_Fundamental_Medio                -0.259433   
Escolaridade_Superior_Pos                      0.259433   

                                LimitedoChequeEspecial  \
Idade                                         0.075883   
RendimentoTotal                               0.858120   
Salario                                       0.888839   
LimitedeCreditoImediato                       0.226990   
LimitedoChequeEspecial                        1.000000   
Escolaridade_Fundamental_Medio               -0.598310   
Escolaridade_Superior_Pos                     0.598310   

                                Escolaridade_Fundamental_Medio  \
Idade                                                -0.110892   
RendimentoTotal                                      -0.396191   
Salario                                              -0.475479   
LimitedeCreditoImediato                              -0.259433   
LimitedoChequeEspecial                               -0.598310   
Escolaridade_Fundamental_Medio                        1.000000   
Escolaridade_Superior_Pos                            -1.000000   

                                Escolaridade_Superior_Pos  
Idade                                            0.110892  
RendimentoTotal                                  0.396191  
Salario                                          0.475479  
LimitedeCreditoImediato                          0.259433  
LimitedoChequeEspecial                           0.598310  
Escolaridade_Fundamental_Medio                  -1.000000  
Escolaridade_Superior_Pos                        1.000000

In [6]:
import statsmodels.formula.api as smf

In [7]:
function = "LimitedoChequeEspecial~Idade+RendimentoTotal+Salario+LimitedeCreditoImediato+Escolaridade_Fundamental_Medio+Escolaridade_Superior_Pos"

In [10]:
model = smf.ols(formula=function, data=limite).fit()
print(model.summary())

                              OLS Regression Results                              
Dep. Variable:     LimitedoChequeEspecial   R-squared:                       0.848
Model:                                OLS   Adj. R-squared:                  0.830
Method:                     Least Squares   F-statistic:                     46.89
Date:                    Wed, 02 Aug 2023   Prob (F-statistic):           4.07e-16
Time:                            15:30:31   Log-Likelihood:                -441.00
No. Observations:                      48   AIC:                             894.0
Df Residuals:                          42   BIC:                             905.2
Df Model:                               5                                         
Covariance Type:                nonrobust                                         
                                     coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------

In [12]:
# Já que temos poucas variáveis, optamos por fazer o método Backward e iremos excluir a idade, uma vez que ficou em 91% (P-Value)

function = "LimitedoChequeEspecial~RendimentoTotal+Salario+LimitedeCreditoImediato+Escolaridade_Fundamental_Medio+Escolaridade_Superior_Pos"
model = smf.ols(formula=function, data=limite).fit()
print(model.summary())

                              OLS Regression Results                              
Dep. Variable:     LimitedoChequeEspecial   R-squared:                       0.848
Model:                                OLS   Adj. R-squared:                  0.834
Method:                     Least Squares   F-statistic:                     59.99
Date:                    Wed, 02 Aug 2023   Prob (F-statistic):           4.92e-17
Time:                            15:33:09   Log-Likelihood:                -441.01
No. Observations:                      48   AIC:                             892.0
Df Residuals:                          43   BIC:                             901.4
Df Model:                               4                                         
Covariance Type:                nonrobust                                         
                                     coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------

In [14]:
# Excluindo do rendimento total que ficou em 18% (P-Value) e -1 é para excluir a variavel Intercept

function = "LimitedoChequeEspecial~Salario+LimitedeCreditoImediato+Escolaridade_Fundamental_Medio+Escolaridade_Superior_Pos -1"
model = smf.ols(formula=function, data=limite).fit()
print(model.summary())

                              OLS Regression Results                              
Dep. Variable:     LimitedoChequeEspecial   R-squared:                       0.842
Model:                                OLS   Adj. R-squared:                  0.831
Method:                     Least Squares   F-statistic:                     77.92
Date:                    Wed, 02 Aug 2023   Prob (F-statistic):           1.23e-17
Time:                            15:35:50   Log-Likelihood:                -442.00
No. Observations:                      48   AIC:                             892.0
Df Residuals:                          44   BIC:                             899.5
Df Model:                               3                                         
Covariance Type:                nonrobust                                         
                                     coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------

In [15]:
# Excluindo a Escolaridade_Superior_Pos, pois ficou em 93%
function = "LimitedoChequeEspecial~Salario+LimitedeCreditoImediato+Escolaridade_Fundamental_Medio -1"
model = smf.ols(formula=function, data=limite).fit()
print(model.summary())

                                   OLS Regression Results                                  
Dep. Variable:     LimitedoChequeEspecial   R-squared (uncentered):                   0.915
Model:                                OLS   Adj. R-squared (uncentered):              0.909
Method:                     Least Squares   F-statistic:                              160.9
Date:                    Wed, 02 Aug 2023   Prob (F-statistic):                    4.57e-24
Time:                            15:37:17   Log-Likelihood:                         -442.01
No. Observations:                      48   AIC:                                      890.0
Df Residuals:                          45   BIC:                                      895.6
Df Model:                               3                                                  
Covariance Type:                nonrobust                                                  
                                     coef    std err          t      P>|t|      

Assim chegamos a 90% de acuracia, alegando que a variavel Limite do Cheque é especial é extremamente afetada pela oscilação do Salário, Limite de Crédito Imediato e Escolaridade (Fundamental/Médio)

In [7]:
#Foram incluídas novas variáveis no Dataset

new_limit = pd.read_excel("Limite_Credito_Dummy.xlsx")
new_limit.head()

Idade  RendimentoTotal  Salario  LimitedeCreditoImediato  \
0     70            10736     5214                      400   
1     68             5000     5000                      380   
2     44             5800     5800                      500   
3     72             4300     4300                     2000   
4     75             4400     4400                     3000   

        Escolaridade Gênero Região  LimitedoChequeEspecial  
0  Fundamental_Medio     RJ      F                     500  
1  Fundamental_Medio     RJ      F                     600  
2  Fundamental_Medio     RJ      F                     800  
3  Fundamental_Medio     RJ      F                    1000  
4  Fundamental_Medio     RJ      F                    1000

In [8]:
new_limit = pd.get_dummies(data=new_limit, columns=["Escolaridade","Gênero","Região"], prefix=["Escolaridade","Gênero","Região"])
new_limit.head()

Idade  RendimentoTotal  Salario  LimitedeCreditoImediato  \
0     70            10736     5214                      400   
1     68             5000     5000                      380   
2     44             5800     5800                      500   
3     72             4300     4300                     2000   
4     75             4400     4400                     3000   

   LimitedoChequeEspecial  Escolaridade_Fundamental_Medio  \
0                     500                               1   
1                     600                               1   
2                     800                               1   
3                    1000                               1   
4                    1000                               1   

   Escolaridade_Superior_Pos  Gênero_RJ  Gênero_SP  Região_F  Região_M  
0                          0          1          0         1         0  
1                          0          1          0         1         0  
2                          0          1          0         1         0  
3                          0          1          0         1         0  
4                          0          1          0         1         0

In [9]:
new_limit.corr()

Idade  RendimentoTotal   Salario  \
Idade                           1.000000         0.239445  0.213186   
RendimentoTotal                 0.239445         1.000000  0.960551   
Salario                         0.213186         0.960551  1.000000   
LimitedeCreditoImediato         0.057183         0.031041  0.081792   
LimitedoChequeEspecial          0.228117         0.858120  0.888839   
Escolaridade_Fundamental_Medio -0.150088        -0.396191 -0.475479   
Escolaridade_Superior_Pos       0.150088         0.396191  0.475479   
Gênero_RJ                      -0.196052        -0.177798 -0.242621   
Gênero_SP                       0.196052         0.177798  0.242621   
Região_F                       -0.121217        -0.406241 -0.424335   
Região_M                        0.121217         0.406241  0.424335   

                                LimitedeCreditoImediato  \
Idade                                          0.057183   
RendimentoTotal                                0.031041   
Salario                                        0.081792   
LimitedeCreditoImediato                        1.000000   
LimitedoChequeEspecial                         0.226990   
Escolaridade_Fundamental_Medio                -0.259433   
Escolaridade_Superior_Pos                      0.259433   
Gênero_RJ                                     -0.001145   
Gênero_SP                                      0.001145   
Região_F                                      -0.473326   
Região_M                                       0.473326   

                                LimitedoChequeEspecial  \
Idade                                         0.228117   
RendimentoTotal                               0.858120   
Salario                                       0.888839   
LimitedeCreditoImediato                       0.226990   
LimitedoChequeEspecial                        1.000000   
Escolaridade_Fundamental_Medio               -0.598310   
Escolaridade_Superior_Pos                     0.598310   
Gênero_RJ                                    -0.312566   
Gênero_SP                                     0.312566   
Região_F                                     -0.591779   
Região_M                                      0.591779   

                                Escolaridade_Fundamental_Medio  \
Idade                                                -0.150088   
RendimentoTotal                                      -0.396191   
Salario                                              -0.475479   
LimitedeCreditoImediato                              -0.259433   
LimitedoChequeEspecial                               -0.598310   
Escolaridade_Fundamental_Medio                        1.000000   
Escolaridade_Superior_Pos                            -1.000000   
Gênero_RJ                                             0.237508   
Gênero_SP                                            -0.237508   
Região_F                                              0.602464   
Região_M                                             -0.602464   

                                Escolaridade_Superior_Pos  Gênero_RJ  \
Idade                                            0.150088  -0.196052   
RendimentoTotal                                  0.396191  -0.177798   
Salario                                          0.475479  -0.242621   
LimitedeCreditoImediato                          0.259433  -0.001145   
LimitedoChequeEspecial                           0.598310  -0.312566   
Escolaridade_Fundamental_Medio                  -1.000000   0.237508   
Escolaridade_Superior_Pos                        1.000000  -0.237508   
Gênero_RJ                                       -0.237508   1.000000   
Gênero_SP                                        0.237508  -1.000000   
Região_F                                        -0.602464  -0.167248   
Região_M                                         0.602464   0.167248   

                                Gênero_SP  Região_F  Região_M  
Idade                            0.196052 -0.121217  0

In [10]:
function = "LimitedoChequeEspecial~Idade+RendimentoTotal+Salario+LimitedeCreditoImediato+Escolaridade_Fundamental_Medio+Escolaridade_Superior_Pos+Gênero_RJ+Gênero_SP+Região_F+Região_M"
model = smf.ols(formula=function , data=new_limit).fit()
print(model.summary())

                              OLS Regression Results                              
Dep. Variable:     LimitedoChequeEspecial   R-squared:                       0.886
Model:                                OLS   Adj. R-squared:                  0.865
Method:                     Least Squares   F-statistic:                     44.20
Date:                    Wed, 02 Aug 2023   Prob (F-statistic):           7.44e-17
Time:                            19:02:26   Log-Likelihood:                -434.21
No. Observations:                      48   AIC:                             884.4
Df Residuals:                          40   BIC:                             899.4
Df Model:                               7                                         
Covariance Type:                nonrobust                                         
                                     coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------

In [14]:
function = "LimitedoChequeEspecial~RendimentoTotal+Salario+LimitedeCreditoImediato+Escolaridade_Superior_Pos+Gênero_RJ+Gênero_SP+Região_F+Região_M"
model = smf.ols(formula=function , data=new_limit).fit()
print(model.summary())

                              OLS Regression Results                              
Dep. Variable:     LimitedoChequeEspecial   R-squared:                       0.885
Model:                                OLS   Adj. R-squared:                  0.869
Method:                     Least Squares   F-statistic:                     52.78
Date:                    Wed, 02 Aug 2023   Prob (F-statistic):           9.97e-18
Time:                            19:05:27   Log-Likelihood:                -434.24
No. Observations:                      48   AIC:                             882.5
Df Residuals:                          41   BIC:                             895.6
Df Model:                               6                                         
Covariance Type:                nonrobust                                         
                                coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------------------------------------------------

In [16]:
function = "LimitedoChequeEspecial~Salario+Gênero_RJ+Gênero_SP+Região_F+Região_M -1"
model = smf.ols(formula=function , data=new_limit).fit()
print(model.summary())

                              OLS Regression Results                              
Dep. Variable:     LimitedoChequeEspecial   R-squared:                       0.879
Model:                                OLS   Adj. R-squared:                  0.871
Method:                     Least Squares   F-statistic:                     106.7
Date:                    Wed, 02 Aug 2023   Prob (F-statistic):           3.24e-20
Time:                            19:06:16   Log-Likelihood:                -435.50
No. Observations:                      48   AIC:                             879.0
Df Residuals:                          44   BIC:                             886.5
Df Model:                               3                                         
Covariance Type:                nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Salario     

Com isso temos uma explicabilidade de 87%, sobre as variaveis explicativas { Salário | Genero (SP/RJ) | Região (F/M)} em função da variação da Target { Limite do Cheque Especial }